graph/patterns/upsert.md

```diff
@@ -0,0 +1,173 @@
+
+This solution allows for existing resources that follow Microsoft Graph conventions for CRUD operations to add `UPSERT` without impacting existing apps or functionality.
+
+Ideally, all new entity types should support an `UPSERT` mechanism, especially if the resource is likely be used in IaC scenarios.
+
+## When to use this pattern
+
+This pattern should be adopted for resources that are managed through infrastructure as code or desired state configuration.
+
+## Issues and considerations
+
+* The addition of this new pattern (with alternate key) does not represent a breaking change.
+However, some API producers may have concerns about accidental usages of this new pattern unwittingly creating many new resources when the intent was an update.
+As a result, API producers can use the `Prefer: idempotent` to require clients to opt-in to the UPSERT behavior.
+* The client-provided alternate key must be immutable after being set. If its value is null then it should be settable as a way to backfill existing resources for use in IaC scenarios.
+* API producers could use `PUT` operations to create or update, but generally this approach is not recommended due to the destructive nature of `PUT`'s replace semantics.
+* API producers could to use `UPSERT` with a primary (client-provided) key and this may be appropriate for some scenarios. However, the recommendation is for resources to support creation using `POST` and a service-generated primary key, for consistency reasons.
+* API producers may annotate entity sets, singletons and collections to indicate that entities can be "upserted". The example below shows this annotation for the `groups` entity set.  
+
+```xml
+<EntitySet Name="groups" EntityType="microsoft.graph.group">
+  <Annotation Term="Org.OData.Capabilities.V1.UpdateRestrictions">
+    <Record>
+      <PropertyValue Property="Upsertable" Bool="true"/>
+    </Record>
+  </Annotation>
+</EntitySet>
+```
+
+## Examples
+
+For these examples we'll use the `group` entity type, which defines both a primary (service-generated) key (`id`) and an alternate (client-provided) key (`uniqueName`).  
+
+```xml
+<EntityType Name="group">
+  <Key>
+    <PropertyRef Name="id"/>
+  </Key>
+  <Property Name="id" Type="Edm.String"/> 
+  <Property Name="uniqueName" Type="Edm.String"/>
+  <Property Name="displayName" Type="Edm.String"/>
+  <Property Name="description" Type="Edm.String"/> 
+  <Annotation Term="OData.Community.Keys.V1.AlternateKeys">
+    <Collection>
+      <Record Type="OData.Community.Keys.V1.AlternateKey">
+        <PropertyValue Property="Key">
+        <Collection>
+            <Record Type="OData.Community.Keys.V1.PropertyRef">
+            <PropertyValue Property="Name" PropertyPath="uniqueName" />
+            </Record>
+        </Collection>
+        </PropertyValue>
+      </Record>
+    </Collection>
+  </Annotation>
+  </Property>
+</EntityType>
+```
+
+### Upserting a record (creation path)
+
```

Comment:
If Upsert behavior is an opt-in based on the Prefer header, then we should probably demonstrate it with an example